src/enzyme_ad/jax/Passes/StablehloOptPatterns.td

```diff
@@ -18,44 +39,48 @@ include "stablehlo/dialect/StablehloOps.td"
 // folding applied after.
 
 // Pattern: op(op(X, cstA), cstB) -> op(X, op(cstA, cstB))
-class BinOpLiftConstantComputation<Op OpTypeA, Op OpTypeB,
+class BinOpLiftConstantComputation<string mnemonic,
+                                   Op OpTypeA, Op OpTypeB,
                                    Op ResOpTypeA = OpTypeA, Op ResOpTypeB = OpTypeB>
-  : Pat<(OpTypeA:$opA
+  : StablehloPatternOp<mnemonic,
+        (OpTypeA:$opA
           (OpTypeB:$opB $lhsB, (StableHLO_ConstantOp:$rhsB $valueB)),
           (StableHLO_ConstantOp:$rhsA $valueA)),
         (ResOpTypeA $lhsB, (ResOpTypeB $rhsB, $rhsA))>;
 
 // Flip the order of the constants, useful for divide and subtract.
 // Pattern: op(op(X, cstA), cstB) -> op(X, op(cstB, cstA))
-class BinOpLiftConstantComputationFlip<Op OpTypeA, Op OpTypeB,
+class BinOpLiftConstantComputationFlip<string mnemonic,
+                                       Op OpTypeA, Op OpTypeB,
                                        Op ResOpTypeA = OpTypeA, Op ResOpTypeB = OpTypeB>
-  : Pat<(OpTypeA:$opA
-          (OpTypeB:$opB $lhsB, (StableHLO_ConstantOp:$rhsB $valueB)),
-          (StableHLO_ConstantOp:$rhsA $valueA)),
-        (ResOpTypeA $lhsB, (ResOpTypeB $rhsA, $rhsB))>;
+  : StablehloPatternOp<mnemonic,
+          (OpTypeA:$opA
+            (OpTypeB:$opB $lhsB, (StableHLO_ConstantOp:$rhsB $valueB)),
+            (StableHLO_ConstantOp:$rhsA $valueA)),
+          (ResOpTypeA $lhsB, (ResOpTypeB $rhsA, $rhsB))>;
 
 // Naming convention: `<RootOp><ChildOp>Const`:
 
 // Pattern: (x / cst_1) * cst_2 -> x * (cst_2 / cst_1)
-def MulDivConst : BinOpLiftConstantComputationFlip<StableHLO_MulOp, StableHLO_DivOp>;
+def MulDivConst : BinOpLiftConstantComputationFlip<"mul_div_const", StableHLO_MulOp, StableHLO_DivOp>;
```

Comment:
How useful is it having a ton of individual patterns as transform ops?

I'm thinking a few possible outcomes here -- we offer individual patterns _and_ grouped pattern ops for transform (i.e. {`mul_div_const`, `mul_mul_const`,...} and `binop_const`) -- or maybe this is something we resolve in the python layer with something like aliases?

I guess the real question is -- is there any issue with exploding the number of transform extension ops we're providing here?

I'd like to expand this to all StableHLO optim patterns so we can think about using some of the upstream ones (and upstreaming new ones) without impacting usability:
https://github.com/openxla/stablehlo/blob/main/stablehlo/transforms/optimization/StablehloAggressiveSimplificationPatterns.td

cc @ftynse @wsmoses

Reply:
So part of the original motivation for having the names was debugging both performance and issues.

For example, one can excise specific patterns to find bugs like https://github.com/EnzymeAD/Enzyme-JAX/pull/424.

For performance it gives a search space since some patterns aren't always beneficial in practice (though from the hlo side it may seem strictly simplifying).

I don't think this means we need to have all things as individual patterns, but we can group things where convenient.

Reply:
The alternative would be to have one `stablehlo_patterns` op that is further parameterized by a list of patterns to include/exclude. But that means having (yet another) registration mechanism for that list whereas now we can use the op registration mechanism. Having logical groupings to reduce clutter makes sense to me as a short-term trade-off. Longer-term, we need to find a good way to manage this with e.g., adding ops on the fly or somehow referring to pattern name/location without creating a whole new op every time.

Reply:
Sounds good, agree. I think a few groupings are probably best in the short term, might remove the individual granularity and just use the `StablehloPatternGroup` from this PR.

I like the file locality of this approach in the meantime to keep things trivial for users contributing patterns / also lets you refer to identifiers in TD files for slightly-better-than-stringly-typed safety. Also will make it really easy to give existing StableHLO patterns bindings in transform dialect